### PR TITLE
Start the runloop asynchronously rather than synchronously during activation

### PR DESCRIPTION
**Pull request by rajsesh-msft**
Our activation code is starting runloop synchornously in the main thread.

This turned out to be bad for some apps.  This change is to do an aysnc task that sets up the runloop.

### CODE REVIEW

**Comment by rajsesh-msft** [comment]
@jaredhms is added to the review. #Closed

File: Frameworks/UIKit/StarboardXaml/ApplicationMain.mm [hunk 1]
```diff
@@ -147,7 +147,7 @@ int ApplicationMainStart(const char* principalName,
     [displayMode _updateDisplaySettings];
 
     UIApplicationMainInit(principalClassName, delegateClassName, defaultOrientation, (int)activationType, activationArgument);
-    ScheduleMainRunLoop();
```
Comment (rajsesh-msft) [inline Frameworks/UIKit/StarboardXaml/ApplicationMain.mm]: >ScheduleMainRunLoop [](start = 4, length = 19)

This was taking > 20s for internal test app.  With async, we take that cost after app activation so PLM doesn't kill us.

File: Frameworks/Foundation/MainDispatcher.mm [hunk 2]
```diff
@@ -85,6 +85,34 @@ void _DispatchMainRunLoopOnUIThread(int signaledEvent) {
 }
 
 /**
+* Private method to create the dispatcher
+*/
+void _ensureDispatcher() {
```
Comment (jaredhms) [inline Frameworks/Foundation/MainDispatcher.mm]: >_ensureDispatcher [](start = 5, length = 17)

nit: should be _PascalCase

**Comment by jaredhms** [comment]
:shipit:

**Comment by msft-Jeyaram** [comment]
:shipit:

File: Frameworks/Foundation/MainDispatcher.mm [hunk 3]
```diff
@@ -85,6 +85,34 @@ void _DispatchMainRunLoopOnUIThread(int signaledEvent) {
 }
 
 /**
+* Private method to create the dispatcher
+*/
+void _EnsureDispatcher() {
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/MainDispatcher.mm]: super-nit: should this be static?

File: Frameworks/Foundation/MainDispatcher.mm [hunk 4]
```diff
@@ -85,6 +85,34 @@ void _DispatchMainRunLoopOnUIThread(int signaledEvent) {
...
+/**
+ * Method that schedules the main runloop asynchronously on the UI thread.
+ * Note: This method always schedules the work asynchronously on the UI thread to let the caller stack unwind.
+ */
+extern "C" void ScheduleMainRunLoopAsync() {
+    auto dispatchCallback = WRLHelpers::MakeAgileCallback<IDispatchedHandler>([]() -> HRESULT {
+        ScheduleMainRunLoop();
+        return S_OK;
+    });
+    _ensureDispatcher();
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/MainDispatcher.mm]: I think one `e` was missed here!

File: Frameworks/Foundation/MainDispatcher.mm [hunk 5]
```diff
@@ -85,6 +85,34 @@ void _DispatchMainRunLoopOnUIThread(int signaledEvent) {
 }
 
 /**
+* Private method to create the dispatcher
+*/
+void _EnsureDispatcher() {
+    if (s_dispatcher == nullptr) {
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/MainDispatcher.mm]: Can we be sure this is thread-safe?

Comment (rajsesh-msft) [inline Frameworks/Foundation/MainDispatcher.mm]: Good point.  This is always initialized in the main thread, so the _EnsureDispatcher in the other thread is unnecessary, I'll remove that.

---
In reply to: [83293021](https://github.com/Microsoft/WinObjC/pull/1142#discussion_r83293021) [](ancestors = 83293021)

**Pull request merged**